### PR TITLE
Created a failing test for two types which refer to each other

### PR DESCRIPTION
**Pull request by Tisten**
TL;DR: and made a fix for the failure

More details:
The old code had inline definitions of operator[], begin() and end() for C++, but if a type which wasn't declared yet were used then it caused a compile error:
```
error C2143: syntax error: missing ';' before '*'
error C4430: missing type specifier - int assumed. Note: C++ does not support default-int
error C2238: unexpected token(s) preceding ';'
```
and even with a forward declare it caused an error:
```
error C2036: 'MyTypeName *': unknown size
```
The C code were fine, but the generated header weren't very readable.

I introduced a macro for the array, and a templated implementation of the C++ case which solves the issue.

### CODE REVIEW

File: src/dl_typelib_write_c_header.cpp [hunk 1]
```diff
@@ -113,6 +113,86 @@ static void dl_context_write_c_header_begin( dl_binary_writer* writer, const cha
...
+									   "#  endif\n"
+									   "\n"
+									   "   // ... DLArray ...\n"
+									   "#  if defined(__cplusplus)\n"
+									   "	  template <typename T>\n"
+									   "	  struct DLArrayStruct\n"
+									   "	  {\n"
+									   "		   inline const T& operator[](uint32_t i) const\n"
+									   "		   {\n"
+									   "		       DL_ASSERT(i < count);\n"
```
[the remark this answers is not in the document]
Comment (wc-duck) [reply src/dl_typelib_write_c_header.cpp:129]: hmmm... not really sure. As you can see here https://github.com/wc-duck/datalibrary/#asserts DL_ASSERT() can be overridden, and I guess that can be useful in this case, but that would also require the conf-header to be included here as well.

File: src/dl_typelib_write_c_header.cpp [hunk 2]
```diff
@@ -113,6 +113,86 @@ static void dl_context_write_c_header_begin( dl_binary_writer* writer, const cha
...
+									   "		       return data[i];\n"
+									   "		   }\n"
+									   "		   inline T& operator[](uint32_t i)\n"
```
Comment (wc-duck) [inline src/dl_typelib_write_c_header.cpp:132]: I'm no fan of switching the type here... what was the reason for changing the API here?

File: src/dl_typelib_write_c_header.cpp [hunk 3]
```diff
@@ -113,6 +113,86 @@ static void dl_context_write_c_header_begin( dl_binary_writer* writer, const cha
...
+									   "\n"
+									   "   // ... DL_ASSERT() ...\n"
+									   "#  if !defined(DL_ASSERT)\n"
+									   "#    define DL_ASSERT(x) // default to non-implemented.\n"
+									   "#  endif\n"
+									   "\n"
+									   "   // ... DLArray ...\n"
+									   "#  if defined(__cplusplus)\n"
+									   "	  template <typename T>\n"
+									   "	  struct DLArrayStruct\n"
```
Comment (wc-duck) [inline src/dl_typelib_write_c_header.cpp:125]: I know naming is DL is not unified across the codebase and that is a problem, but the public api is always lower_case_underscore.
I would go with 'dl_array' I think...

File: src/dl_typelib_write_c_header.cpp [hunk 4]
```diff
@@ -113,6 +113,86 @@ static void dl_context_write_c_header_begin( dl_binary_writer* writer, const cha
...
+									   "		       return data;\n"
+									   "		   }\n"
+									   "		   inline const char** end()\n"
+									   "		   {\n"
+									   "		       return data + count;\n"
+									   "		   }\n"
+									   "		   const char** data;\n"
+									   "		   uint32_t count;\n"
+									   "	   };\n"
+									   "#	 define DLArray(type) DLArrayStruct<type>\n"
```
Comment (wc-duck) [inline src/dl_typelib_write_c_header.cpp:187]: macros is always UPPER_CASE_UNDERSCORE... DL_DECLARE_ARRAY() maybe?

**Review by wc-duck — CHANGES_REQUESTED** [review]

File: src/dl_typelib_write_c_header.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (Tisten) [reply src/dl_typelib_write_c_header.cpp:129]: That would turn dl_config.h from a private header to a public one, which isn't what I was after. This assert is meant to check usage of the generated types, not to validate the internals of DL.

File: src/dl_typelib_write_c_header.cpp (hunk 2 above)
Comment (Tisten) [reply src/dl_typelib_write_c_header.cpp:132]: I will change it back. I wasn't even aware that I changed it. I based the code on the fact that the array `count` is `uint32_t` and not size_t. The "size" of the array is set to 16 bytes though, so I'm not sure what the purpose of not using size_t is?

File: src/dl_typelib_write_c_header.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (Tisten) [reply src/dl_typelib_write_c_header.cpp]: Changed it to DL_DATA_ASSERT for now

File: src/dl_typelib_write_c_header.cpp (hunk 2 above)
Comment (Tisten) [reply src/dl_typelib_write_c_header.cpp]: If you would have used pragma pack the size would have been 12... That could have been nice as well...

**Pull request merged**